### PR TITLE
db-write hook

### PR DESCRIPTION
**Pull request by rustyrussell**
It's a bit limited (see final commits for caveats) but it works.

See #2400

### CODE REVIEW

**Comment by rustyrussell** [comment]
Hmm, needs sqlite3 v3.14 or above.  Not sure what that would break if we include it, might have to make hook optional?  Yech...

**Comment by cdecker** [comment]
I was expecting to see an implementation of `sqlite3_expanded_sql` took me while to figure out that it is actually a native `sqlite3` function. Maybe we can shim this? This hook seems far too useful not to exist. Notice though that it makes switching DBs even harder since plugins now may depend on the `sqlite3` dialect being used.

File: wallet/db.h [hunk 1]
```diff
@@ -75,6 +78,35 @@ void db_set_intvar(struct db *db, char *varname, s64 val);
...
+ */
+#define db_select_prepare(db, query) \
+	db_select_prepare_(__FILE__ ":" stringify(__LINE__), db, query)
+sqlite3_stmt *db_select_prepare_(const char *location,
+				 struct db *db, const char *query);
+
+/**
+ * db_select_step -- iterate through db results.
+ *
+ * Returns false and frees stmt if we've we've reached end, otherwise
```
Comment (niftynei) [inline wallet/db.h]:
```suggestion
 * Returns false and frees stmt if we've reached end, otherwise
```

File: doc/PLUGINS.md [hunk 2]
```diff
@@ -270,5 +268,25 @@ gossiped list of known addresses. In particular this means that the port for
...
 reconnections.
 
+#### `db_write`
+
+This hook is called whenever a change is about to be committed to the database.
+It is currently extremely restricted:
+
+1. a plugin registering for this hook should not perform anything that may cause
+   a db operation in response (pretty much, anything but logging).
+2. a plugin registering for this hook should not register for other hooks or
```
Comment (niftynei) [inline doc/PLUGINS.md]: should we canonize this by returning an error if this is broken?

File: ccan/ccan/io/io.h [hunk 3]
```diff
@@ -722,6 +722,35 @@ bool io_plan_out_started(const struct io_conn *conn);
...
+ * @exclusive: whether to be exclusive or not
+ *
+ * If any io_conn is set exclusive, then no non-exclusive io_conn (or
+ * io_listener) will be serviced by io_loop().  If it's a io_duplex io_conn(),
+ * then io_conn_exclusive() makes the read-side exclusive; io_conn_out_exclusive()
+ * makes the write-side exclusive.
+ *
+ * This allows you to temporarily service only one (or several) fds.
+ * For example, you might want to flush out one io_conn and not
+ * receive any new connections or read any otherninput.
```
Comment (niftynei) [inline ccan/ccan/io/io.h]:
```suggestion
 * receive any new connections or read any other input.
```

File: ccan/ccan/io/io.h [hunk 4]
```diff
@@ -722,6 +722,35 @@ bool io_plan_out_started(const struct io_conn *conn);
...
+ *
+ * Returns true of there any exclusive io_conn remain, otherwise false.
```
Comment (niftynei) [inline ccan/ccan/io/io.h]:
```suggestion
 * Returns true if any exclusive io_conn remain, otherwise false.
```

**Comment by rustyrussell** [comment]
> I was expecting to see an implementation of `sqlite3_expanded_sql` took me while to figure out that it is actually a native `sqlite3` function. Maybe we can shim this? This hook seems far too useful not to exist. Notice though that it makes switching DBs even harder since plugins now may depend on the `sqlite3` dialect being used.

I think we can try to use the (now obsolete) sqlite3 tracing feature to simulate this in older sqlite3.  Let me try.

**Comment by rustyrussell** [comment]
> really like the `db_select_step` improvements.
> 
> may be worth noting that in theory, your 'backup' copy may be one command set ahead of the actual on disk database if it crashes/fails during write... tho i guess that's actually the point eh. :)

Yes, but in this case both dbs are "valid".  You could naively copy the whole db in the plugin every startup, but better would be a separate "db_init" hook, which would use https://www.sqlite.org/c3ref/c_fcntl_begin_atomic_write.html#sqlitefcntldataversion, and give plugin the version startup time.

If the plugin also has an sqlite3 db, that's a trivial check, and if not, it can fall back to copy.

File: wallet/db.h (hunk 1 above)
Comment (rustyrussell) [reply wallet/db.h]: :+1:

**Comment by rustyrussell** [comment]
OK, put in backwards compat shim, rebased.

**Comment by rustyrussell** [comment]
Ack 1d776eb4c1e264913715dbd1bcbb6ba534686bb7

Let's get it in so we can make progress :)

**Comment by Coinomatron** [comment]
Updated to the latest commits, it's not clear to me how this backup function works. Can someone give a brief explanation?

**Comment by cdecker** [comment]
> Updated to the latest commits, it's not clear to me how this backup function works. Can someone give a brief explanation?

It's not a backup mechanism on its own, it is the infrastructure necessary to write backup plugins on top.

**Comment by Coinomatron** [comment]
Ah, thanks.

**Comment by cdecker** [comment]
No problem, we'll get a move on and get some example implementations soon :-)

**Comment by Coinomatron** [comment]
Any updates on this?

**Comment by ZmnSCPxj** [comment]
My last information is that @jb55 was working on some backup system that implements this.

**Comment by jb55** [comment]
oh yeah I should start that

**Comment by jb55** [comment]
hmm now that you reminded me, I didn't go that route as I was going to see how the postgresql backend played out. if you could use a postgresql backend that replicates offsite, that perhaps could solve the problem as well? cc @cdecker 

although that's not very user friendly, so perhaps a plugin is still in order...

that makes me wonder how this interacts with the multi-db abstraction now... :thinking: I haven't looked at the code in either case yet

**Pull request merged**